compiler-core/src/language_server/code_action.rs

```diff
@@ -909,3 +909,449 @@ pub fn code_action_add_missing_patterns(
+        Self {
+            module,
+            params,
+            line_numbers,
+            edits: vec![],
+            qualified_constructor: None,
+            should_stop: false,
+            second_pass: false,
+        }
+    }
+
+    pub fn code_actions(mut self) -> Vec<CodeAction> {
+        self.visit_typed_module(&self.module.ast);
+        self.second_pass = true;
+        self.visit_typed_module(&self.module.ast);
+        if self.edits.is_empty() {
+            return vec![];
+        }
+        let mut action = Vec::with_capacity(1);
+        CodeActionBuilder::new("Convert to unqualified import")
+            .kind(CodeActionKind::REFACTOR)
+            .changes(self.params.text_document.uri.clone(), self.edits)
+            .preferred(false)
+            .push_to(&mut action);
+        action
+    }
+
+    fn get_module_import(&mut self, module_name: &EcoString) -> Option<&'a ast::Import<EcoString>> {
+        self.module
+            .ast
+            .definitions
+            .iter()
+            .find_map(|def| match def {
+                ast::Definition::Import(import) if import.module == *module_name => Some(import),
+                _ => None,
+            })
+    }
+
+    fn find_in_code(
+        &self,
+        start: usize,
+        end: usize,
+        predicate: impl Fn(&str) -> Option<usize>,
+    ) -> Option<usize> {
+        self.module
+            .code
+            .get(start..end)
+            .and_then(predicate)
+            .map(|pos| start + pos)
+    }
+
+    fn find_trailing_comma_before_closing_brace(
+        &self,
+        import_start: usize,
+        import_end: usize,
+    ) -> Option<usize> {
+        let import_code = self.module.code.get(import_start..import_end)?;
+        let closing_brace_pos = import_code.rfind('}')?;
+
+        let bytes = import_code.as_bytes();
+        let mut pos = closing_brace_pos;
+        while pos > 0 {
+            pos -= 1;
+            let c = (*bytes.get(pos)?) as char;
+            if c.is_whitespace() {
+                continue;
+            }
+            if c == ',' {
+                return Some(import_start + pos);
+            }
+            break;
+        }
+        None
+    }
+
+    fn edit_import(&mut self, name: String) {
+        let qualified_constructor = self
+            .qualified_constructor
+            .as_ref()
+            .expect("import should be set");
+        let import_start = qualified_constructor.import.location.start as usize;
+        let import_end = qualified_constructor.import.location.end as usize;
+
+        // Determine if the import has braces
+        let has_brace = self
+            .find_in_code(import_start, import_end, |s| s.rfind('}'))
+            .is_some();
+
+        let trailing_comma_pos = if has_brace {
+            self.find_trailing_comma_before_closing_brace(import_start, import_end)
+        } else {
+            None
+        };
+
+        let (insert_pos, new_text) = match (
```

Comment:
Favour guards or some other structure over matching on 2 bools in a tuple as these patters are hard to understand.